### PR TITLE
Archive contract artifacts

### PR DESCRIPTION
**Pull request by mslipper**
To demonstrate how we can make our deployments more modular, this PR proposes archiving smart contract artifacts as tarballs that get uploaded to GCS. This  allows deployment tools to use precompiled artifacts rather than generating them on-the fly.

The archives are named after the hash of all Solidity files in the contracts-bedrock folder, including those in `lib/`, plus some additional metadata files like `foundry.toml` and `semver-lock.json`. See `calculate-checksum.sh` for details on how the algorithm works. I'm open to feedback around what should make up the checksum.

Since the atifacts are content-addressable, this PR also updates the CI pipeline to download the artifacts from GCS prior to running `pnpm monorepo`. When the Solidity codebase doesn't change, this allows the `pnpm monorepo` job to skip compiling Solidity altogether. While this won't work as well when we're actively modifying the Solidity codebase, since the hash will change, it does provide a modest speedup in CI.

### CODE REVIEW

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
I haven't seen this actually run, but the design makes sense

File: .circleci/config.yml [hunk 1]
```diff
@@ -1461,6 +1465,24 @@ jobs:
           working_directory: ./packages/contracts-bedrock
       - notify-failures-on-develop
 
+  publish-contract-artifacts:
```
Comment (tynes) [inline .circleci/config.yml:1468]: One thought is we should be fork friendly, ie not break in forks by:
- using top level variables that allows a small diff for forks
- putting things behind an `if repo name != ethereum-optimism/optimism: skip()`

Comment (mslipper) [reply .circleci/config.yml:1468]: Good point - I can add that. Right now it'll show as "unauthorized" if a fork tries to run it.

**Comment by mslipper** [comment]
>  I haven't seen this actually run, but the design makes sense.

It ran on this PR if you'd like to check it out!

**Comment by tynes** [comment]
> > I haven't seen this actually run, but the design makes sense.
> 
> It ran on this PR if you'd like to check it out!

Can confirm, thank you!

https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/62533/workflows/b97cbf77-c3d6-40d6-ab53-fc5758e40214/jobs/2606204

All looks good to me!

File: .circleci/config.yml (hunk 1 above)
Comment (mslipper) [reply .circleci/config.yml:1468]: On second thought, I'll merge this now and add this in separately. I'm not sure there's a way to skip jobs if they need a context; I'll have to look into it.

**Pull request merged**